fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -1,69 +1,398 @@
+            log,
+            data: None,
+        };
+        Ok(r)
+    } else {
+        let swap_fee = amount;
+        let effective_amount = Uint128::zero();
+
+        let log = vec![
+            log("action", "reverse_swap"),
+            log("rid", rid),
+            log("to", to),
+            log("from", from),
+            log("origin_tx_hash", origin_tx_hash),
+            log("amount", effective_amount),
+            log("swap_fee", swap_fee),
+        ];
+
+        let r = HandleResponse {
+            messages: vec![],
+            log,
+            data: None,
+        };
+        Ok(r)
+    }
+}
+
+/* ***************************************************
+ * *****************    Helpers      *****************
+ * ***************************************************/
+
+fn amount_from_funds(funds: &Vec<Coin>) -> Uint128 {
+    // TODO(LR) return proper error
+    if funds.len() == 1
+    /* && funds[0].denom == "fet" */
+    {
+        // TODO(LR) does cosmwas allows sending multiple funds of the same token
+        funds[0].amount
+    } else {
+        Uint128::zero()
+    }
+}
+
+fn send_tokens<A: Api>(
+    api: &A,
+    env: &Env,
+    state: &State,
+    to_address: &CanonicalAddr,
+    amount: Uint128,
+    action: &str,
+) -> HandleResult {
+    let from_address = &env.contract.address;
+    let from_human = api.human_address(&from_address)?;
+    let to_human = api.human_address(to_address)?;
+    let log = vec![log("action", action), log("to", to_human.as_str())];
+    let coin = Coin {
+        amount,
+        denom: state.denom.clone(),
+    };
+
+    let r = HandleResponse {
+        messages: vec![CosmosMsg::Bank(BankMsg::Send {
+            from_address: from_human,
+            to_address: to_human,
+            amount: vec![coin],
+        })],
+        log,
+        data: None,
+    };
+    Ok(r)
+}
+
+/* ***************************************************
+ * ***************    Verifiers      *****************
+ * ***************************************************/
+
+fn verify_tx_relay_eon(eon: u64, state: &State) -> HandleResult {
+    if eon != state.relay_eon {
+        Err(StdError::generic_err(
+            "Tx doesn't belong to current relayEon",
+        ))
+    } else {
+        Ok(HandleResponse::default())
+    }
+}
+
+fn verify_not_paused(env: &Env, state: &State) -> HandleResult {
+    if env.block.height >= state.paused_since_block {
+        Err(StdError::generic_err(format!(
+            "Contract has been paused {}",
+            state.paused_since_block
+        )))
+    } else {
+        Ok(HandleResponse::default())
+    }
+}
+
+fn verify_swap_amount(amount: Uint128, state: &State) -> HandleResult {
+    if amount < state.lower_swap_limit {
+        Err(StdError::generic_err("Amount bellow lower limit"))
+    } else if amount > state.upper_swap_limit {
+        Err(StdError::generic_err("Amount exceeds upper limit"))
+    } else if (state.supply + amount) > state.cap {
```

Comment:
Just comment: solely due to performance reasons, it wuld be better to:
* **either** relocate the `else if (state.supply + amount) > state.cap` check out of this verif. function in to `try_swap(...)` function, in order to perform the `state.supply + amount` calculation just once, not twice,
* **OR**, alternativelly, calculate the new `state.supply` once here in this function and return it as return value

Reply:
@pbukva I agree, I saw that in your implementation you followed suggestion 1, still I wanted to have separation of concerns where the `verify_swap_amount` will do all the verifications, but I guess performance is more important for smart contract.
(for information, the second suggestion breaks the verifiy-ers pattern by returning a value)

Reply:
You have several options here:
1. either to leave it as-is,
2. calculate `state.supply + amount` before calling this function and pass it in as second input paraneter,
3. calculate `state.supply + amount` and find a way how to pass it out as return value, or by mutable reference input parameter ...

Reply:
Yes, I understood that Peter :)
I will change it to compute the sum only once and do the check outside `verify_swap_amount`